Describe `ImageTexture`, `Image` creation and usage

Xrayez opened the pull request:
Closes #20972.
Helps #41106, #42346, #38112, #23148, #16678, #17848.

Having this documented, I'd prefer #39396 or #42653 to get merged, as it's mostly documentation issue by now, see #24222 for instance.

Linking to the [Importing images](https://docs.godotengine.org/en/3.2/getting_started/workflow/assets/importing_images.html) docs as well.

Commit:
Describe `ImageTexture`, `Image` creation and usage

Xrayez commented on doc/classes/Image.xml:
```diff
@@ -4,10 +4,12 @@
-		[b]Note:[/b] The maximum image size is 16384×16384 pixels due to graphics hardware limitations. Larger images will fail to import.
+		Native image datatype. Contains image data which can be converted to an [ImageTexture] and provides commonly used [i]image processing[/i] methods. The maximum width and height for an [Image] are [constant MAX_WIDTH] and [constant MAX_HEIGHT].
+		An [Image] cannot be assigned to a [code]texture[/code] property of an object directly (such as [Sprite2D]), and has to be converted manually to an [ImageTexture] first.
+		[b]Note:[/b] The maximum image size is 16384×16384 pixels due to graphics hardware limitations. Larger images may fail to import.
```
This is not true in `master` now, as it depends on `MAX_PIXELS` constant, but true for `3.2`, will have to be updated (need to expose `MAX_PIXELS` constant first). The limits are still legit for textures.

xarses commented on doc/classes/ImageTexture.xml:
```diff
@@ -4,10 +4,36 @@
+		texture.create_from_image(image)
+		$Sprite.texture = texture
+		[/codeblock]
+		[b]Warning:[/b] Always prefer to load imported textures over loading them from within the filesystem dynamically, as it may not work in exported projects:
```
I've seen this warning multiple times from many sources, but still don't understand the why's of it. I also continue to not understand how to know when an texture is "imported" properly v.s. just present. There is also an implication that "imported" objects are inside the project folder before the game is exported but I've not seen this clearly defined. Lastly what should someone do if they truly intend for an image/texture to be loaded from `user://`?

Xrayez replied:
In theory this should be documented at [`Import process`](https://docs.godotengine.org/en/stable/getting_started/workflow/assets/import_process.html) page describing the reasons why assets like images have to be imported in Godot.

By default, any foreign file format is not recognized by Godot, so it will be ignored and not visible throughout the "FileSystem" dock (which looks into `res://` path, which happens to be a project's root on your filesystem, or the contents of `game.pck` when you export a project).

On the high-level, files like `*.png` are recognized by Godot, but they have to be made compatible with the existing datatypes first, like `Image` or `Texture`. Therefore, those foreign files have to be converted first, aka imported as `Resource`s. Those imported resources get saved into the `.import` folder which is hidden in Godot's UI.

The problem is that there may be different ways to convert such files, so the "Import" dock gives you the ability to configure the default import procedure for each file, as described in https://github.com/godotengine/godot/issues/20972#issuecomment-645514487.

Now, if you want to actually interact with those imported resources, loading them by the "original path" is enough:

```gdscript
var texture = load("res://icon.png")
```

What happens is that the `ResourceLoader` takes the original path ("res://icon.png"), looks whether such a file was previously imported (looks into `.import` directory), and retrieves the resource for you, so it doesn't really load the original file anymore.

Now, the problem with such approach is that this is limited to the project's virtual resource filesystem ("res://"). If you need to load any external files outside of "res://" in exported projects, this will fail, because all of the game contents get packed into `game.pck`. Methods like `Image.load()` are not even aware of this virtual filesystem in Godot and rely on the typical OS-specific filesystem methods for loading data (using the internal `FileAccess` class, aka `File` in GDScript), and not GDScript's `load()`. The reason it works in the editor is because those image files are still there on your filesystem. In theory, you can also disable automatic re-import, delete those original files outside of the editor, and still able to run your game. 🙂

> Lastly what should someone do if they truly intend for an image/texture to be loaded from `user://`?

For best results, you have to replicate the same import process manually at run-time, because most of the import methods are only available in the editor. Yes, I know, this can be a big limitation in your project if your game takes advantage of user-generated content. And the most practical and feasible way to "import" textures at run-time is by loading an `Image` and creating an `ImageTexture` from that. 🙂 

The procedure would be the same as in the first snippet in this PR, but use "user://" instead. You have to clearly define what content would be able to be loaded externally and design your game architecture accordingly.

xarses replied:
thanks this was very helpful

Xrayez commented:
Updated documentation to provide some rationale as to why textures are better to be loaded with `load` rather than `Image.load()`, and suggest that `Image.load()` is only safe to use in the context of the editor and for the use case of loading external images (`user://`) in exported projects.

Xrayez commented:
Updated upon discussion in https://github.com/godotengine/godot/issues/42346#issuecomment-701549568.

Xrayez commented:
Updated upon https://github.com/godotengine/godot/issues/42346#issuecomment-702263401. Added error messages for `ImageTexture.update()` function and `create_from_image()`.

`ImageTexture.set_data()` in 3.2 is `ImageTexture.update()` in 4.0.

xarses commented:
LGTM, the changes resolve the problems I ran into.

akien-mga commented on doc/classes/ImageTexture.xml:
```diff
@@ -4,10 +4,38 @@
+		[b]Note:[/b] The maximum texture size is 16384×16384 pixels due to graphics hardware limitations.
 	</description>
 	<tutorials>
+		<link title="Importing images">https://docs.godotengine.org/en/latest/getting_started/workflow/assets/importing_images.html</link>
```
It moved: https://docs.godotengine.org/en/latest/tutorials/assets_pipeline/importing_images.html

akien-mga commented on doc/classes/ImageTexture.xml:
```diff
@@ -4,10 +4,38 @@
+		[/codeblock]
+		This is because [method Image.load] relies on OS-specific file access methods that bypass the virtual filesystem implemented in the engine to load imported resources, so [method Image.load] should only be used in the editor or in cases when you need to load external images at run-time, such as images located at the [code]user://[/code] directory.
+		If you'd still like to load an image file just like any other [Resource], import it as an [Image] resource instead, and then load it normally using the [method @GDScript.load] method:
+		[codeblock]
+		var image = load("res://icon.png")
+		assert(image is Image)
+		assert(not (image is ImageTexture))
+		[/codeblock]
```
I wouldn't include this codeblock as it doesn't tell users *how* to import the file as `Image`, it just shows a way to do validation. IMO it's better left out.

Xrayez replied:
The code is more like documentation here, rather than an example of class usage, but ok I'll remove this.

akien-mga commented on doc/classes/ImageTexture.xml:
```diff
@@ -4,10 +4,38 @@
+		But do note that the image data can still be retrieved from an imported texture as well using the [method Texture2D.get_data] method, which returns a copy of the data:
+		[codeblock]
+		var texture = load("res://icon.png")
+		var image = texture.get_data()
+		assert(image is Image)
```
```suggestion
		var image : Image = texture.get_data()
```
IMO type hints are a better way to convey type information than asserts.

Xrayez replied:
I just recall you weren't happy with including type hints in documentation. And I think `assert()` works better for detecting potential bugs if users copy-paste code snippets. This is more like doctest in Python, but that's up for a proposal!

akien-mga replied:
Well `Texture2D.get_data()`'s return type is `Image`, so the `assert()` here is not useful, as the type cannot be something else than `Image`. `image.is_empty()` might be closer to the kind of sanity checks that users could do, but that's also outside the scope of this documentation IMO.

Type hints are not used currently in the docs, but if they make sense in specific code blocks, they can be used.

Otherwise another way is:
```
# Texture2D.get_data() returns an Image type.
var image = texture.get_data()
```

I won't die on this hill though :)

akien-mga commented on doc/classes/ImageTexture.xml:
```diff
@@ -4,10 +4,38 @@
+		$Sprite2D.texture = texture
+		[/codeblock]
+		This way, textures can be created at run-time by loading images both from within the editor and externally.
+		[b]Warning:[/b] Prefer to load imported textures over loading them from within the filesystem dynamically, as it may not work in exported projects:
```
I got confused by this as the different between `Prefer to load imported textures over loading them from within the filesystem dynamically` is not immediately obvious.

I'd suggest maybe:
```suggestion
		[b]Warning:[/b] Prefer to load imported textures with [method @GDScript.load] over loading them from within the filesystem dynamically with [method Image.load], as it may not work in exported projects:
```

akien-mga commented on doc/classes/ImageTexture.xml:
```diff
@@ -4,10 +4,38 @@
+		var texture = load("res://icon.png")
+		$Sprite2D.texture = texture
+		[/codeblock]
+		This is because [method Image.load] relies on OS-specific file access methods that bypass the virtual filesystem implemented in the engine to load imported resources, so [method Image.load] should only be used in the editor or in cases when you need to load external images at run-time, such as images located at the [code]user://[/code] directory.
```
While I agree that using imported StreamTextures should be recommended, it's not clear to me why/how `Image.load()` would fail due to OS-specific file access methods.

Do you mean that it will fail in exported projects in `res://` because the actual PNG file is not in the PCK? If so that's not really OS/file access specific, but due to the file being imported as a StreamTexture and thus not compatible with using as Image.

Xrayez replied:
Simplified the rationale that images are imported as StreamTextures. Some information is available in `Image` already.

akien-mga commented:
Thanks!

akien-mga commented:
This is not trivial to cherry-pick as the API for textures changed, and this impacts some of the written documentation. Can you make a dedicated PR for `3.2`?
